### PR TITLE
Extract generic UnpublishView and add unpublish action to snippets with DraftStateMixin

### PR DESCRIPTION
**Pull request by laymonage**
-   [x] Do the tests still pass?
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

**Please describe additional details for testing this change**.

- Create a snippet with DraftStateMixin applied
- Publish the snippet
- Open the edit view and click on the unpublish action menu

### CODE REVIEW

File: wagtail/snippets/tests/test_snippets.py [hunk 1]
```diff
@@ -1641,6 +1692,122 @@ def test_get_after_publish_and_save_draft(self):
...
+
+        # Should be redirected to explorer page
+        self.assertRedirects(
+            response, reverse("wagtailsnippets_tests_draftstatemodel:list")
+        )
+
+        # Check that the page was unpublished
+        self.assertFalse(DraftStateModel.objects.get(pk=self.snippet.pk).live)
+
+        # Check that the page_unpublished signal was fired
```
Comment (jacobtoppm) [inline wagtail/snippets/tests/test_snippets.py:1767]: I thought it wouldn't be the `page_unpublished`?

**Review by jacobtoppm — APPROVED** [review]
The code looks good aside from some minor nitpicks

File: wagtail/snippets/tests/test_snippets.py (hunk 1 above)
Comment (laymonage) [reply wagtail/snippets/tests/test_snippets.py]: Oops yes, copy-paste leftover, it should just be `unpublished`.

File: wagtail/snippets/tests/test_snippets.py [hunk 2]
```diff
@@ -1641,6 +1692,122 @@ def test_get_after_publish_and_save_draft(self):
...
+        # Check that the user received a 404 response
+        self.assertEqual(response.status_code, 404)
+
+    def test_unpublish_view_bad_permissions(self):
+        """
+        This tests that the unpublish view doesn't allow users without unpublish permissions
+        """
+        # Remove privileges from user
+        self.user.is_superuser = False
+        self.user.user_permissions.add(
```
[the remark this answers is not in the document]
Comment (laymonage) [reply wagtail/snippets/tests/test_snippets.py:1735]: It's only used in one test in this class, and there are other test classes with the same pattern in this file (and this test was copied almost verbatim from the unpublish tests for pages), so maybe we could just leave that in?

[the remark this answers is not in the document]
Comment (jacobtoppm) [reply wagtail/snippets/tests/test_snippets.py:1735]: Eh, fair enough. I don't think it's a blocker.

**Pull request merged**